File: day-2/main.py
def is_safe_report(report):
    # Convert the report string into a list of integers
    levels = list(map(int, report.split()))
    
    # Check if the levels are strictly increasing or decreasing
    increasing = all(1 <= levels[i+1] - levels[i] <= 3 for i in range(len(levels) - 1))
    decreasing = all(1 <= levels[i] - levels[i+1] <= 3 for i in range(len(levels) - 1))
    
    return increasing or decreasing

def can_be_safe_with_removal(levels):
    # Iterate through each level and check if removing it makes the report safe
    for i in range(len(levels)):
        modified_levels = levels[:i] + levels[i+1:]
        if is_safe_report(' '.join(map(str, modified_levels))):
            return True
    return False

def count_safe_reports_part1(data):
    reports = data.strip().split('\n')
    safe_count = 0
    
    # Iterate through each report and check if it is safe
    for report in reports:
        if is_safe_report(report):
            print(f"Safe report (Part 1): {report}")
            safe_count += 1
        else:
            print(f"Unsafe report (Part 1): {report}")
    
    return safe_count

def count_safe_reports_part2(data):
    reports = data.strip().split('\n')
    safe_count = 0
    
    # Iterate through each report and check if it is safe or can be made safe by removing one level
    for report in reports:
        levels = list(map(int, report.split()))
        if is_safe_report(report):
            print(f"Safe report (Part 2): {report}")
            safe_count += 1
        elif can_be_safe_with_removal(levels):
            print(f"Report made safe by removal (Part 2): {report}")
            safe_count += 1
        else:
            print(f"Unsafe report (Part 2): {report}")
    
    return safe_count

if __name__ == "__main__":
    # Read input data from file
    with open('input.txt', 'r') as file:
        input_data = file.read()
    
    # Count safe reports for Part 1
    part1_safe_count = count_safe_reports_part1(input_data)
    # Count safe reports for Part 2
    part2_safe_count = count_safe_reports_part2(input_data)
    
    # Print results
    print(f"Part 1: {part1_safe_count} safe reports")
    print(f"Part 2: {part2_safe_count} safe reports")

File: day-2/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func isSafeReport(report string) bool {
	levels := parseLevels(report)
	increasing := true
	decreasing := true

	for i := 0; i < len(levels)-1; i++ {
		diff := levels[i+1] - levels[i]
		if diff < 1 || diff > 3 {
			increasing = false
		}
		if diff > -1 || diff < -3 {
			decreasing = false
		}
	}

	return increasing || decreasing
}

func canBeSafeWithRemoval(levels []int) bool {
	for i := range levels {
		modifiedLevels := append([]int{}, levels[:i]...)
		modifiedLevels = append(modifiedLevels, levels[i+1:]...)
		if isSafeReport(joinLevels(modifiedLevels)) {
			return true
		}
	}
	return false
}

func countSafeReportsPart1(data string) int {
	reports := strings.Split(strings.TrimSpace(data), "\n")
	safeCount := 0
	for _, report := range reports {
		if isSafeReport(report) {
			safeCount++
		}
	}
	return safeCount
}

func countSafeReportsPart2(data string) int {
	reports := strings.Split(strings.TrimSpace(data), "\n")
	safeCount := 0
	for _, report := range reports {
		levels := parseLevels(report)
		if isSafeReport(report) {
			safeCount++
		} else if canBeSafeWithRemoval(levels) {
			safeCount++
		}
	}
	return safeCount
}

func parseLevels(report string) []int {
	parts := strings.Fields(report)
	levels := make([]int, len(parts))
	for i, part := range parts {
		levels[i], _ = strconv.Atoi(part)
	}
	return levels
}

func joinLevels(levels []int) string {
	strLevels := make([]string, len(levels))
	for i, level := range levels {
		strLevels[i] = strconv.Itoa(level)
	}
	return strings.Join(strLevels, " ")
}

func main() {
	file, err := os.Open("input.txt")
	if err != nil {
		fmt.Println("Error opening file:", err)
		return
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	var inputData string
	for scanner.Scan() {
		inputData += scanner.Text() + "\n"
	}

	part1SafeCount := countSafeReportsPart1(inputData)
	part2SafeCount := countSafeReportsPart2(inputData)

	fmt.Printf("Part 1: %d safe reports\n", part1SafeCount)
	fmt.Printf("Part 2: %d safe reports\n", part2SafeCount)
}
